### DISCUSSION TITLE
defining zod schemas using goja and go

### DISCUSSION THREAD

**Discussion by ymolists** [dop251/goja#615] [Q&A]
Hi Folks

I am trying to load zod using goja so i can define my types using zod from go.

would you kindly help me figure how i can import the released zod modules and use them from typescript ? 

here is what i did so far
1) i am able to compile a typescript file by embedding (using go embed) the latest typescript compiler in go. yay !
2) now i am stuck on figuring what is the minimal zod file(s) i need and how to import it 
i created a new npm package and added zod as a dependency i then copied what i had in my node_modules/zod folder to zod-3.23.8/ there is a bunch of files[2] not sure which ones i need to import and which ones i can discard.

here is how i want to import zod in my ts file . 

[1]
```ts 
import { z } from "zod";

const mySchema = z
  .object({
    myString: z.string().min(5),
    myUnion: z.union([z.number(), z.boolean()]),
  })
  .describe("My neat object schema");
```

[2]
```
zod-3.23.8/
zod-3.23.8/package.json
zod-3.23.8/index.d.ts
zod-3.23.8/lib
zod-3.23.8/lib/errors.js
zod-3.23.8/lib/external.d.ts
zod-3.23.8/lib/external.js
zod-3.23.8/lib/index.d.ts
zod-3.23.8/lib/ZodError.js
zod-3.23.8/lib/__tests__
zod-3.23.8/lib/__tests__/Mocker.js
zod-3.23.8/lib/__tests__/Mocker.d.ts
zod-3.23.8/lib/index.js
zod-3.23.8/lib/types.js
zod-3.23.8/lib/errors.d.ts
zod-3.23.8/lib/locales
zod-3.23.8/lib/locales/en.js
zod-3.23.8/lib/locales/en.d.ts
zod-3.23.8/lib/ZodError.d.ts
zod-3.23.8/lib/helpers
zod-3.23.8/lib/helpers/typeAliases.d.ts
zod-3.23.8/lib/helpers/parseUtil.js
zod-3.23.8/lib/helpers/partialUtil.js
zod-3.23.8/lib/helpers/errorUtil.d.ts
zod-3.23.8/lib/helpers/enumUtil.js
zod-3.23.8/lib/helpers/util.js
zod-3.23.8/lib/helpers/partialUtil.d.ts
zod-3.23.8/lib/helpers/util.d.ts
zod-3.23.8/lib/helpers/enumUtil.d.ts
zod-3.23.8/lib/helpers/errorUtil.js
zod-3.23.8/lib/helpers/parseUtil.d.ts
zod-3.23.8/lib/helpers/typeAliases.js
zod-3.23.8/lib/index.umd.js
zod-3.23.8/lib/index.mjs
zod-3.23.8/lib/types.d.ts
zod-3.23.8/lib/benchmarks
zod-3.23.8/lib/benchmarks/ipv4.d.ts
zod-3.23.8/lib/benchmarks/datetime.d.ts
zod-3.23.8/lib/benchmarks/discriminatedUnion.d.ts
zod-3.23.8/lib/benchmarks/discriminatedUnion.js
zod-3.23.8/lib/benchmarks/object.js
zod-3.23.8/lib/benchmarks/union.d.ts
zod-3.23.8/lib/benchmarks/index.d.ts
zod-3.23.8/lib/benchmarks/datetime.js
zod-3.23.8/lib/benchmarks/realworld.d.ts
zod-3.23.8/lib/benchmarks/ipv4.js
zod-3.23.8/lib/benchmarks/index.js
zod-3.23.8/lib/benchmarks/string.d.ts
zod-3.23.8/lib/benchmarks/object.d.ts
zod-3.23.8/lib/benchmarks/realworld.js
zod-3.23.8/lib/benchmarks/primitives.js
zod-3.23.8/lib/benchmarks/string.js
zod-3.23.8/lib/benchmarks/union.js
zod-3.23.8/lib/benchmarks/primitives.d.ts
zod-3.23.8/README.md
zod-3.23.8/LICENSE
```
 

Regards

**Comment by dop251** [comment]
There is no ESM support yet, but you can use CJS modules using `require` from [goja_nodejs](https://github.com/dop251/goja_nodejs)

**Reply by ymolists** [reply]
Oh great ! thank you for jumping on this !!!

Do you think doing a require from `zod-3.23.8/lib/index.js` would be enough ? The plan is to hook in the require function and when i detect the actual module is `zod` i replace that with `zod-3.23.8/lib/index.js` in go. 

also would you mind clarifying which ones are `cjs` and wjich ones are `esm` from the above zod files list ? i am bundling these files with my go binary using embed so i will make sure they are not part of the binary. 

i am mostly a go developer so there is a skill issue here with all things zod/ts/js.

Regards
